plugins/mirror/bazel.go

```diff
@@ -40,6 +40,17 @@ func (a *Artifact) AppendURL(url string) {
 	a.rule.Attr("urls").(*build.ListExpr).List = append(list, &build.StringExpr{Value: url})
 }
 
+func (a *Artifact) RemoveURL(url string) {
```

Comment:
How about a `RemoveURLs(urls []string) {}` function, where you do something like this (did not compile it):


```go
func (a *Artifact) RemoveURL(urls []string) {
    urlsToRemove := map[string]struct{}{}
    for i, _ := range urls {
        urlsToRemove[urls[i]] = struct{}
    }
    
    var list ListExpr
    for i, urlVal := range a.rule.Attr("urls").(*build.ListExpr).List {
        if _, exists := urlToRemove[urlVal.Value]; !exists {
            append(list, &build.StringExpr{Value: urlVal.Value})
        } 
    }
    a.rule.Attr("urls").(*build.ListExpr).List = list
}
```

It has the advantage of being less tricky to understand while being save for corner cases.

If you first fetch in `RemoveStaleDownloadURLS` all urls which you need to get rid of, and then pass them all to this function, you end up with a much lower runtime complexity, and it is still easy to follow.

Reply:
Great point, although the runtime of the `http.Head` will far outweigh the reduced complexity, besides that there are around five URLs per artifact right now, which when it's run periodically will be reduced further.  :slightly_smiling_face:

Reply:
I am more interested in worst case scenarios. I can easily construct a WORKSPACE file which will have funny effects. I prefer writing the code save. Note that not necessarily malicious attacks can cause issues. Chaining automated processes canas well lead to bad situations.

It is easy to rewrite and also clear to follow: 1) detect bad content, 2) remove it. The current way on how to remove elements from the list is not really easier from a cognitive perspective.